### ISSUE TITLE
[Discovery] Custom Fonts: support `woff2`

### ISSUE THREAD

**Issue by miina** [swissspidy/web-stories-wp#10020]
## Feature Description
Currently we're using `opentype.js` for getting the information about the font from the font file URL, however, this library doesn't support `woff2`.

There are some libraries that support `woff2`: `opentype` and `fontkit`. Both of these mention also having browser support, however, apparently both modules require `fs` module by default in various places which causes the editor to crash.

There are a few GitHub issues in the repos of both modules stating the concern of not being able to use these in the browser with a bunch of suggested solutions for `webpack`, however, many of those didn't work when trying / were quite outdated and caused more issues.

We should take another look to figure out a way to support `woff2`.

**Comment by miina** [comment]
@swissspidy FYI added this issue but I'm not sure if it'll be Infra / Prometheus or shared.
When I was looking into this, it seemed like getting the woff2-supporting libraries to work in browser would need diving into `webpack` configuration.

**Comment by merapi** [comment]
I've spent some time on this, and I've got the following conclusions:

### opentype.js WOFF2 support
Adding support for WOFF2 to opentype.js is equal to a complete refactor of the [`parseBuffer`](https://github.com/opentypejs/opentype.js/blob/88f7c1755ceec3326cdac4de53bc5ec89d38a615/src/opentype.js#L171) part because it's not as easy as [some people think](https://github.com/opentypejs/opentype.js/issues/183#issuecomment-207958637) (just throw in a 🥦  decompressor), probably that's why other people tried and failed to achieve it since 2016.
Fundamentally [WOFF](https://www.w3.org/Submission/WOFF/) and [WOFF2](https://www.w3.org/TR/WOFF2/#FileStructure) files have a different structure.

TLDR: WOFF has more fixed structs, each table is compressed on its own and it uses fixed offsets defined in the table directory for each table vs WOFF2 that has no fixed offsets, and the data for the font tables is compressed in a single data stream.

As a result, the flow of the program is very different and the current code is hard to adapt - it has to be rewritten. While exploring this I've successfully implemented the parsing of the table directory for WOFF2 in opentype.js but after that, a lot more work is needed.

### Using another library that has no browser support (although they claim it has)
I saw that some people had success with replacing `fs` with ~`null` but only if the functionality used by them didn't need the `fs` module, otherwise it will not work, the same for the `zlib` module (used for 🥦 ).

I'm looking right now at `fontkit` because it has some potential.

**Comment by merapi** [comment]
Interesting find in `opentype` dependency `brotli`:
https://unpkg.com/brotli@1.3.2/dec/dictionary-browser.js

That `var fs = require('fs');` seems to do nothing here, except breaking the browser usage 😉 

The fixed version is unreleased (and repo ☠️ ): https://github.com/foliojs/brotli.js/issues/30

**Comment by merapi** [comment]
Bottom line: most of those libraries aren't written with the browser in mind and would require some modifications to work either by switching out the dependencies (forking or patching with upstream PRs?) or implementing woff2 support.

We don't need most of the features that libraries like `fontkit` offer (consider bundle size), we don't need writing to font files (includes Brotli compression), we just need to read some of the font metrics. So it makes some sense to develop our tailor-made solution, but it will take time (days). On the other hand, woff2 fonts can be easily converted to woff1 by the end-user.

Depending on the demand, we should decide what to do next.

**Comment by miina** [comment]
Let's not close this issue before we actually decide what to do and if we need a follow-up implementation issue or not :)
Moving to Review instead.

**Comment by miina** [comment]
Note: @bmattb is also checking if this feature (woff2) is a priority for the users of the MVP version, this might also help us decide if we want to spend time on a custom solution for supporting it.

**Comment by miina** [comment]
New issue added with P4 priority: https://github.com/GoogleForCreators/web-stories-wp/issues/10440

As of today, it looks like NC might not have this as a requirement, closing the discovery issue. Thanks for looking into this, Marcin!
